pantheon/src/main/java/tech/pegasys/pantheon/cli/custom/EnodeToURIPropertyConverter.java

```diff
@@ -0,0 +1,98 @@
+  private static final String IPV4_PATTERN =
+      "(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}";
+  private static final String IPV6_PATTERN = "(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}";
+  private static final String IPV6_COMPACT_PATTERN =
+      "((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)::((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)";
+  private static final String DISCOVERY_PORT_PATTERN = "\\?discport=(?<discovery>\\d+)";
+
+  private static final String ENODE_URL_PATTERN =
+      "enode://(?<nodeId>\\w+)@(?<ip>" + IP_REPLACE_MARKER + "):(?<listening>\\d+)";
+
+  @Override
+  public URI convert(final String value) throws IllegalArgumentException {
+    if (value == null || "".equals(value)) {
+      throw new IllegalArgumentException("Can't convert null/empty string to EnodeURLProperty.");
+    }
+
+    final boolean containsDiscoveryPort = value.contains("discport");
+    final boolean isIPV4 = Pattern.compile(".*" + IPV4_PATTERN + ".*").matcher(value).matches();
+    final boolean isIPV6 = Pattern.compile(".*" + IPV6_PATTERN + ".*").matcher(value).matches();
+    final boolean isIPV6Compact =
+        Pattern.compile(".*" + IPV6_COMPACT_PATTERN + ".*").matcher(value).matches();
+
+    String pattern = ENODE_URL_PATTERN;
+    if (isIPV4) {
+      pattern = pattern.replace(IP_REPLACE_MARKER, IPV4_PATTERN);
+    } else if (isIPV6) {
+      pattern = pattern.replace(IP_REPLACE_MARKER, IPV6_PATTERN);
+    } else if (isIPV6Compact) {
+      pattern = pattern.replace(IP_REPLACE_MARKER, IPV6_COMPACT_PATTERN);
+    } else {
+      throw new IllegalArgumentException("Invalid enode URL IP format.");
+    }
+
+    if (containsDiscoveryPort) {
+      pattern += DISCOVERY_PORT_PATTERN;
+    }
+    if (isIPV6) {
+      pattern = pattern.replace(IP_REPLACE_MARKER, IPV6_PATTERN);
+    } else {
+      pattern = pattern.replace(IP_REPLACE_MARKER, IPV4_PATTERN);
+    }
+
+    final Matcher matcher = Pattern.compile(pattern).matcher(value);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException(
+          "Invalid enode URL syntax. Enode URL should have the following format 'enode://<node_id>@<ip>:<listening_port>[?discport=<discovery_port>]'.");
+    }
+
+    final String nodeId = matcher.group("nodeId");
+    final String ip = matcher.group("ip");
+    final Integer listeningPort = getAndValidatePort(matcher, "listening");
+
+    if (containsDiscoveryPort(value)) {
+      final Integer discoveryPort = getAndValidatePort(matcher, "discovery");
+      return URI.create(
+          String.format("enode://%s@%s:%d?discport=%d", nodeId, ip, listeningPort, discoveryPort));
+    } else {
+      return URI.create(String.format("enode://%s@%s:%d", nodeId, ip, listeningPort));
+    }
+  }
+
+  private Integer getAndValidatePort(final Matcher matcher, final String portName) {
+    int port = Integer.valueOf(matcher.group(portName));
+    if (port < 0 || port > 65535) {
+      throw new IllegalArgumentException(
```

Comment:
should we have consts for these magic numbers?

Reply:
My bad! I thought they were quite straightforward. I'll add a constant :)

Reply:
Even better! We already have a utility for that :)

`NetworkUtility.isValidPort(port)`